File: src/neml2/models/solid_mechanics/traction_separation_law/BilinearTraction.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "neml2/models/solid_mechanics/traction_separation_law/BilinearTraction.h"
#include "neml2/tensors/Scalar.h"
#include "neml2/tensors/Vec.h"
#include "neml2/tensors/functions/where.h"

namespace neml2
{
register_NEML2_object(BilinearTraction);

OptionSet
BilinearTraction::expected_options()
{
  OptionSet options = Model::expected_options();
  options.doc() =
      "Bilinear cohesive-zone traction with internal damage state. Computes the bilinear damage "
      "variable from \\f$ (\\delta_m, \\delta_c, \\delta_f) \\f$ (effective separation, critical "
      "/ damage-onset separation, full / failure separation), caps it for irreversibility against "
      "the previous-step value (auto-declared via `history_name`), and assembles \\f$ T_n = "
      "K(1-d)\\delta_n^+ + K\\delta_n^-,\\;T_{si} = K(1-d)\\delta_{si} \\f$. Damage is exposed "
      "as a secondary output for diagnostics; the irreversibility cap is internal to this model "
      "and does not require an external `IrreversibleScalar`.";

  options.add_input("effective_separation", "Effective separation");
  options.add_input("normal_separation",
                    "Normal separation (typically the Macaulay-positive part of the normal jump)");
  options.add_input("normal_penetration",
                    "Optional normal penetration (typically the Macaulay-negative part of the "
                    "normal jump). When supplied, K times this is added to the normal traction "
                    "as a penalty term resisting interpenetration.");
  options.add_input("tangential_separation_1", "First tangential separation");
  options.add_input("tangential_separation_2", "Second tangential separation");
  options.add_output("traction", "Traction Vec");
  options.add_output("damage", "Damage scalar (current step, irreversibility-capped)");
  options.add_parameter<Scalar>("penalty_stiffness", "Penalty stiffness");
  options.add_parameter<Scalar>("critical_separation",
                                "Critical (damage-onset) separation. May be wired to an upstream "
                                "`CamanhoDavilaCriticalSeparation` (nonlinear-capable).");
  options.add_parameter<Scalar>("full_separation",
                                "Full (failure) separation. May be wired to an upstream "
                                "`BenzeggaghKenaneFullSeparation` or `PowerLawFullSeparation` "
                                "(nonlinear-capable).");

  return options;
}

BilinearTraction::BilinearTraction(const OptionSet & options)
  : Model(options),
    _to(declare_output_variable<Vec>("traction")),
    _d(declare_output_variable<Scalar>("damage")),
    _d_old(declare_input_variable<Scalar>(history_name(_d.name(), /*nstep=*/1))),
    _delta_m(declare_input_variable<Scalar>("effective_separation")),
    _delta_init(declare_parameter<Scalar>("delta_c",
                                          "critical_separation",
                                          /*allow_nonlinear=*/true)),
    _delta_final(declare_parameter<Scalar>("delta_f",
                                           "full_separation",
                                           /*allow_nonlinear=*/true)),
    _dn_sep(declare_input_variable<Scalar>("normal_separation")),
    _dn_pen(options.user_specified("normal_penetration")
                ? &declare_input_variable<Scalar>("normal_penetration")
                : nullptr),
    _ds1(declare_input_variable<Scalar>("tangential_separation_1")),
    _ds2(declare_input_variable<Scalar>("tangential_separation_2")),
    _K(declare_parameter<Scalar>("K", "penalty_stiffness", /*allow_nonlinear=*/false))
{
}

void
BilinearTraction::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  const auto one = Scalar::ones_like(_delta_m());
  const auto zero = Scalar::zeros_like(_delta_m());

  // -------- Bilinear damage: trial value from (delta_m, delta_c, delta_f). Note that
  // _delta_init and _delta_final are nonlinear-capable parameters declared with
  // allow_nonlinear=true, so they are referenced directly (no operator() like input variables).
  // Detached masks for the where()-conditions (TorchScript can't capture grad-tracking masks).
  const auto df_minus_di = _delta_final - _delta_init;
  const auto df_minus_di_pos = (df_minus_di > 0.0).detach();
  const auto safe_df_minus_di = neml2::where(df_minus_di_pos, df_minus_di, one);
  const auto bilinear_d =
      _delta_final * (_delta_m() - _delta_init) / (_delta_m() * safe_df_minus_di);
  const auto dm_lt_init = (_delta_m() < _delta_init).detach();
  const auto dm_lt_final = (_delta_m() < _delta_final).detach();
  const auto dm_gt_init = (_delta_m() > _delta_init).detach();
  const auto interior = (dm_gt_init && dm_lt_final).detach();
  const auto d_trial = neml2::where(dm_lt_init, zero, neml2::where(dm_lt_final, bilinear_d, one));

  // -------- Irreversibility cap: damage = max(d_trial, d_old).
  const auto advance = (d_trial > _d_old()).detach();
  const auto d = neml2::where(advance, d_trial, _d_old());

  // -------- Active / inactive traction assembly.
  const auto active_scale = _K * (1.0 - d);

  if (out)
  {
    auto T_n = active_scale * _dn_sep();
    if (_dn_pen)
      T_n = T_n + _K * (*_dn_pen)();
    const auto T_s1 = active_scale * _ds1();
    const auto T_s2 = active_scale * _ds2();
    _to = Vec::fill(T_n, T_s1, T_s2);
    _d = d;
  }

  if (!dout_din)
    return;

  // -------- Jacobians.
  // d(d_trial) partials in the interior (zero elsewhere via the `interior` mask):
  //   d/d(delta_m) =  delta_f * delta_c / (delta_m^2 (delta_f - delta_c))
  //   d/d(delta_c) =  delta_f (delta_m - delta_f) / (delta_m (delta_f - delta_c)^2)
  //   d/d(delta_f) = -delta_c (delta_m - delta_c) / (delta_m (delta_f - delta_c)^2)
  const auto inv_dm = 1.0 / _delta_m();
  const auto inv_diff = 1.0 / safe_df_minus_di;
  const auto inv_diff_sq = inv_diff * inv_diff;

  const auto dt_ddm_int = _delta_final * _delta_init * inv_dm * inv_dm * inv_diff;
  const auto dt_ddm = neml2::where(interior, dt_ddm_int, zero);

  // Irreversibility freezes partials when d_trial doesn't advance:
  const auto dd_ddm = neml2::where(advance, dt_ddm, zero);
  const auto dd_dd_old = neml2::where(advance, zero, one);

  // d(damage)/...
  _d.d(_delta_m) = dd_ddm;
  _d.d(_d_old) = dd_dd_old;

  // d(traction)/... routes through (1) the active_scale = K(1-d) factor on
  // (δ_n^sep, δ_s1, δ_s2), (2) the K factor on δ_n^pen for T_n (when supplied), and (3) the
  // per-component multipliers for direct partials.
  const auto neg_K = -_K;
  // dT/d(damage) = -K * (δ_n^sep, δ_s1, δ_s2). Then dT/d(scalar driving damage) =
  // dT/d(damage) * dd/d(scalar).
  const auto dT_dd_n = neg_K * _dn_sep();
  const auto dT_dd_s1 = neg_K * _ds1();
  const auto dT_dd_s2 = neg_K * _ds2();

  _to.d(_delta_m) = Vec::fill(dT_dd_n * dd_ddm, dT_dd_s1 * dd_ddm, dT_dd_s2 * dd_ddm);
  _to.d(_d_old) = Vec::fill(dT_dd_n * dd_dd_old, dT_dd_s1 * dd_dd_old, dT_dd_s2 * dd_dd_old);

  // Nonlinear-parameter Jacobians: only emit when the user wired them to an upstream variable
  // (allow_nonlinear=true). When supplied as literal numbers, nl_param() returns nullptr.
  // Lookup keys are the internal first-arg names of declare_parameter (delta_c, delta_f).
  if (const auto * const dc = nl_param("delta_c"))
  {
    const auto dt_dinit_int = _delta_final * (_delta_m() - _delta_final) * inv_dm * inv_diff_sq;
    const auto dt_dinit = neml2::where(interior, dt_dinit_int, zero);
    const auto dd_dinit = neml2::where(advance, dt_dinit, zero);
    _d.d(*dc) = dd_dinit;
    _to.d(*dc) = Vec::fill(dT_dd_n * dd_dinit, dT_dd_s1 * dd_dinit, dT_dd_s2 * dd_dinit);
  }
  if (const auto * const df = nl_param("delta_f"))
  {
    const auto dt_dfinal_int = -_delta_init * (_delta_m() - _delta_init) * inv_dm * inv_diff_sq;
    const auto dt_dfinal = neml2::where(interior, dt_dfinal_int, zero);
    const auto dd_dfinal = neml2::where(advance, dt_dfinal, zero);
    _d.d(*df) = dd_dfinal;
    _to.d(*df) = Vec::fill(dT_dd_n * dd_dfinal, dT_dd_s1 * dd_dfinal, dT_dd_s2 * dd_dfinal);
  }

  // Direct partials w.r.t. the per-component jumps (d held fixed there, since the cap is on
  // damage, not on the jumps themselves).
  _to.d(_dn_sep) = Vec::fill(active_scale, zero, zero);
  if (_dn_pen)
    _to.d(*_dn_pen) = Vec::fill(Scalar(_K), zero, zero);
  _to.d(_ds1) = Vec::fill(zero, active_scale, zero);
  _to.d(_ds2) = Vec::fill(zero, zero, active_scale);
}
} // namespace neml2

File: neml2/models/solid_mechanics/traction_separation_law/BilinearTraction.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Python-native mirror of the C++ ``BilinearTraction`` model."""

from __future__ import annotations

from ....chain_rule import ChainRuleAction, ChainRuleDict
from ....factory import register_native
from ....model import Model
from ....schema import HitSchema, derived_input, input, output, parameter
from ....types import Scalar, Vec, gt, lt, vec_from_scalars, where


@register_native("BilinearTraction")
class BilinearTraction(Model):
    r"""Bilinear cohesive-zone traction with internal damage state. Computes the
    bilinear damage variable from $(\delta_m, \delta_c, \delta_f)$
    (effective separation, critical / damage-onset separation, full / failure
    separation), caps it for irreversibility against the previous-step value
    (auto-declared via `history_name`), and assembles
    $T_n = K(1-d)\delta_n^+ + K\delta_n^-,\;T_{si} = K(1-d)\delta_{si}$. Damage is
    exposed as a secondary output for diagnostics; the irreversibility cap is
    internal to this model and does not require an external
    `IrreversibleScalar`.
    """

    hit = HitSchema(
        input("effective_separation", Scalar, "Effective separation"),
        input(
            "normal_separation",
            Scalar,
            "Normal separation (typically the Macaulay-positive part of the normal jump)",
        ),
        input(
            "normal_penetration",
            Scalar,
            "Optional normal penetration (typically the Macaulay-negative part of the "
            "normal jump). When supplied, K times this is added to the normal traction "
            "as a penalty term resisting interpenetration.",
            default=None,
            attr="_dn_pen_name",
        ),
        input("tangential_separation_1", Scalar, "First tangential separation"),
        input("tangential_separation_2", Scalar, "Second tangential separation"),
        output("traction", Vec, "Traction Vec"),
        output(
            "damage",
            Scalar,
            "Damage scalar (current step, irreversibility-capped)",
        ),
        # Previous-step damage is auto-declared by appending the ``~1`` history
        # suffix to the resolved ``damage`` output name (matches the C++
        # ``history_name(_d.name(), /*nstep=*/1)`` plumbing). No HIT knob.
        derived_input("damage", Scalar, attr="_d_old_name", suffix="~1"),
        parameter(
            "penalty_stiffness",
            Scalar,
            "Penalty stiffness",
            attr="K",
        ),
        parameter(
            "critical_separation",
            Scalar,
            "Critical (damage-onset) separation. May be wired to an upstream "
            "`CamanhoDavilaCriticalSeparation` (nonlinear-capable).",
            attr="delta_c",
            allow_nonlinear=True,
        ),
        parameter(
            "full_separation",
            Scalar,
            "Full (failure) separation. May be wired to an upstream "
            "`BenzeggaghKenaneFullSeparation` or `PowerLawFullSeparation` "
            "(nonlinear-capable).",
            attr="delta_f",
            allow_nonlinear=True,
        ),
    )

    # ``from_hit`` auto-declares the three Scalar parameters via
    # ``declare_typed_parameter``. The optional ``normal_penetration`` input
    # resolves to a HIT name (string) when the user supplies it, ``None``
    # otherwise; the ``derived_input("damage", ..., suffix="~1")`` resolves
    # ``self._d_old_name`` to ``<resolved damage>~1``.
    K: Scalar
    delta_c: Scalar
    delta_f: Scalar
    _dn_pen_name: str | None
    _d_old_name: str

    def forward(  # type: ignore[override]
        self,
        *args,
        v: ChainRuleDict | None = None,
        **_,
    ):
        # ``input_spec`` order: structural inputs first (declaration order from
        # the schema, minus optional inputs that HIT omitted), then promoted
        # nl-parameters appended in declaration order. Split accordingly.
        names = list(self.input_spec)
        if len(args) != len(names):
            raise AssertionError(
                f"BilinearTraction.forward: got {len(args)} args, expected {len(names)}"
            )
        n_nl = len(self._nl_params)
        n_struct = len(names) - n_nl
        inputs, nl_params = args[:n_struct], args[n_struct:]
        struct_names = names[:n_struct]
        bound = dict(zip(struct_names, inputs, strict=True))

        # ``input_spec`` already holds the *resolved* (post-rename) names, so
        # walk the rename map in reverse to pick out each canonical role.
        renames = getattr(self, "_var_renames", {})

        def _resolved(canon: str) -> str:
            return renames.get(canon, canon)

        delta_m = bound[_resolved("effective_separation")]
        dn_sep = bound[_resolved("normal_separation")]
        ds1 = bound[_resolved("tangential_separation_1")]
        ds2 = bound[_resolved("tangential_separation_2")]
        d_old = bound[self._d_old_name]
        dn_pen_name = self._dn_pen_name
        dn_pen = bound[dn_pen_name] if dn_pen_name is not None and dn_pen_name in bound else None

        # Nonlinear-capable parameters: read via ``_get_param`` so the same
        # code path works whether ``delta_c`` / ``delta_f`` are static
        # parameters or promoted to runtime nl inputs (mode 3/4).
        # ``penalty_stiffness`` is always static (allow_nonlinear=False).
        # The ``_get_param`` lookup key is the schema ``attr`` (which is the
        # ``ctor_name`` used by ``declare_typed_parameter`` in pass 2).
        K = self._get_param("K", nl_params, Scalar)
        delta_c = self._get_param("delta_c", nl_params, Scalar)
        delta_f = self._get_param("delta_f", nl_params, Scalar)

        # -------- Bilinear damage trial value from (delta_m, delta_c, delta_f).
        # Detached masks for ``where()`` (matches C++ ``.detach()``).
        one = Scalar.from_value(1.0, like=delta_m)
        zero = Scalar.from_value(0.0, like=delta_m)
        df_minus_di = delta_f - delta_c
        # ``safe_df_minus_di`` prevents 0/0 at the degenerate ``delta_c ==
        # delta_f`` edge before the ``where`` mask zeros the contribution.
        df_minus_di_pos = gt(df_minus_di, 0.0)
        safe_df_minus_di = where(df_minus_di_pos, df_minus_di, one)
        bilinear_d = delta_f * (delta_m - delta_c) / (delta_m * safe_df_minus_di)
        dm_lt_init = lt(delta_m, delta_c)
        dm_lt_final = lt(delta_m, delta_f)
        # Interior: delta_c < delta_m < delta_f.
        dm_gt_init_data = (delta_m.data > delta_c.data).detach()
        dm_lt_final_data = (delta_m.data < delta_f.data).detach()
        interior_data = (dm_gt_init_data & dm_lt_final_data).detach()
        interior = Scalar(interior_data, sub_batch_ndim=delta_m.sub_batch_ndim)
        d_trial = where(dm_lt_init, zero, where(dm_lt_final, bilinear_d, one))

        # -------- Irreversibility cap: damage = max(d_trial, d_old).
        advance = gt(d_trial, d_old)
        d = where(advance, d_trial, d_old)

        # -------- Assemble traction.
        active_scale = K * (1.0 - d)
        T_n = active_scale * dn_sep
        if dn_pen is not None:
            T_n = T_n + K * dn_pen
        T_s1 = active_scale * ds1
        T_s2 = active_scale * ds2
        T = vec_from_scalars(T_n, T_s1, T_s2)

        if v is None:
            if dn_pen is None:
                # Two outputs: (traction, damage).
                return T, d
            return T, d

        # -------- D-062 pushforward.
        #
        # Trial-value partials in the interior (zero outside):
        #   d(d_t)/d(delta_m) =  delta_f * delta_c / (delta_m^2 (delta_f - delta_c))
        #   d(d_t)/d(delta_c) =  delta_f * (delta_m - delta_f) / (delta_m * (delta_f - delta_c)^2)
        #   d(d_t)/d(delta_f) = -delta_c * (delta_m - delta_c) / (delta_m * (delta_f - delta_c)^2)
        # Irreversibility cap freezes them when d_trial does not advance.
        inv_dm = 1.0 / delta_m
        inv_diff = 1.0 / safe_df_minus_di
        inv_diff_sq = inv_diff * inv_diff

        dt_ddm_int = delta_f * delta_c * inv_dm * inv_dm * inv_diff
        dt_ddm = where(interior, dt_ddm_int, zero)
        # After max(d_t, d_old): partial collapses to 0 on the frozen branch.
        dd_ddm = where(advance, dt_ddm, zero)
        dd_dd_old = where(advance, zero, one)

        # ``-K`` prefactor of d(T)/d(damage); the per-component Vec coefficient
        # is ``(-K * dn_sep, -K * ds1, -K * ds2)`` (zero in the penetration
        # column because the K*dn_pen penalty does not depend on damage).
        neg_K = -K
        dT_dd_n = neg_K * dn_sep
        dT_dd_s1 = neg_K * ds1
        dT_dd_s2 = neg_K * ds2

        # ----- Actions for the damage output. Inputs not listed (dn_sep,
        # dn_pen, ds1, ds2) push forward to structural zero.
        damage_actions: dict[str, ChainRuleAction] = {
            "effective_separation": lambda V, c=dd_ddm: c * V,
            self._d_old_name: lambda V, c=dd_dd_old: c * V,
        }

        # ----- Actions for the traction output. Each direct-jump partial is a
        # Vec with the active_scale (or K) on one component and zero on the
        # other two; damage-mediated partials route through ``-K*[jumps] *
        # dd/d(scalar)``.
        def _traction_via_damage(dd_dscalar: Scalar):
            # d(T)/d(scalar via damage) = (-K * jumps) * dd/d(scalar)
            def action(V: Scalar) -> Vec:
                Vs = V
                return vec_from_scalars(
                    dT_dd_n * dd_dscalar * Vs,
                    dT_dd_s1 * dd_dscalar * Vs,
                    dT_dd_s2 * dd_dscalar * Vs,
                )

            return action

        def _dn_sep_action(V: Scalar) -> Vec:
            Vs = V
            return vec_from_scalars(active_scale * Vs, zero * Vs, zero * Vs)

        def _ds1_action(V: Scalar) -> Vec:
            Vs = V
            return vec_from_scalars(zero * Vs, active_scale * Vs, zero * Vs)

        def _ds2_action(V: Scalar) -> Vec:
            Vs = V
            return vec_from_scalars(zero * Vs, zero * Vs, active_scale * Vs)

        def _dn_pen_action(V: Scalar) -> Vec:
            Vs = V
            return vec_from_scalars(K * Vs, zero * Vs, zero * Vs)

        traction_actions: dict[str, ChainRuleAction] = {
            "effective_separation": _traction_via_damage(dd_ddm),
            self._d_old_name: _traction_via_damage(dd_dd_old),
            "normal_separation": _dn_sep_action,
            "tangential_separation_1": _ds1_action,
            "tangential_separation_2": _ds2_action,
        }
        if dn_pen_name is not None:
            traction_actions[dn_pen_name] = _dn_pen_action

        # ----- Nonlinear-parameter contributions (delta_c, delta_f) — only
        # emit when the user wired them to an upstream variable (allow_nonlinear
        # promoted them into the input set). When supplied as literals the
        # promoted-input name is absent from ``v`` and the action is silently
        # skipped by ``apply_chain_rule``.
        delta_c_nlp = self._nl_params.get("delta_c")
        if delta_c_nlp is not None:
            dt_dinit_int = delta_f * (delta_m - delta_f) * inv_dm * inv_diff_sq
            dt_dinit = where(interior, dt_dinit_int, zero)
            dd_dinit = where(advance, dt_dinit, zero)
            damage_actions[delta_c_nlp.input_name] = lambda V, c=dd_dinit: c * V
            traction_actions[delta_c_nlp.input_name] = _traction_via_damage(dd_dinit)

        delta_f_nlp = self._nl_params.get("delta_f")
        if delta_f_nlp is not None:
            dt_dfinal_int = -delta_c * (delta_m - delta_c) * inv_dm * inv_diff_sq
            dt_dfinal = where(interior, dt_dfinal_int, zero)
            dd_dfinal = where(advance, dt_dfinal, zero)
            damage_actions[delta_f_nlp.input_name] = lambda V, c=dd_dfinal: c * V
            traction_actions[delta_f_nlp.input_name] = _traction_via_damage(dd_dfinal)

        v_T = self.apply_chain_rule(v, "traction", traction_actions, output=T)
        v_d = self.apply_chain_rule(v, "damage", damage_actions, output=d)
        return T, d, {**v_T, **v_d}


__all__ = ["BilinearTraction"]
